code/iaas/model/src/main/java/io/cattle/platform/core/cleanup/TableCleanup.java

```diff
@@ -249,81 +269,115 @@ private void cleanupLabelTables(Date cutoff) {
         for (CleanableTable table : tables) {
-            Field<Long> id = table.idField;
-            Field<Date> remove = table.removeField;
+            if (table.referenceCheckOnly) {
+                // This table is just for reference checking when cleaning other tables. Don't attempt to delete any rows from this table
+                continue;
+            }
 
-            ResultQuery<Record1<Long>> ids = create()
-                    .select(id)
+            Field<Long> idField = table.idField;
+            Field<Date> removeField = table.removeField;
+            ResultQuery<Record1<Long>> idsToDeleteQuery = create()
+                    .select(idField)
                     .from(table.table)
-                    .where(remove.lt(cutoffTime))
+                    .where(removeField.lt(cutoffTime)
+                    .and(idField.gt(DSL.param(ID_GREATER_THAN_PARAM, 0L))))
+                    .orderBy(idField)
                     .limit(QUERY_LIMIT_ROWS.getValue());
+           Param<?> idGreaterThanParam = idsToDeleteQuery.getParam(ID_GREATER_THAN_PARAM);
 
             table.clearRowCounts();
-            Result<Record1<Long>> toDelete;
-            HashSet<Long> idsToFix = new HashSet<Long>();
-            HashSet<Long> idsToFilter = new HashSet<Long>();
-            ResultQuery<Record1<Long>> filterIds;
-            Result<Record1<Long>> toFilter;
-
-            if ((toDelete = ids.fetch()).size() > 0) {
-                List<ForeignKey<?, ?>> keys = getReferencesFrom(table, tables);
-                for (ForeignKey<?, ?> key : keys) {
-                    if (key.getFields().size() > 1) {
-                        log.error("Composite foreign key filtering unsupported");
-                    }
-                }
-                for (ForeignKey<?, ?> key : keys) {
-                    Table<?> referencingTable = key.getTable();
-                    Field<Long> foreignKeyField = (Field<Long>) key.getFields().get(0);
-
-                    filterIds = create().selectDistinct(foreignKeyField).from(referencingTable);
+            Result<Record1<Long>> idsToDeleteResult;
+            HashSet<Long> idsThatCantBeDeleted = new HashSet<Long>();
+            HashSet<Long> idsReferencedInOtherTables = new HashSet<Long>();
+            ResultQuery<Record1<Long>> idsReferencedQuery;
+
+            idsToDeleteResult = idsToDeleteQuery.fetch();
+            if (idsToDeleteResult.size() == 0) {
+                // No rows to remove, move on to next table
+                continue;
+            }
 
-                    if ((toFilter = filterIds.fetch()).size() > 0) {
-                        for (Record1<Long> record : toFilter) {
-                            idsToFilter.add(record.value1());
-                        }
-                    }
+            // Build a set of ids referenced by other tables
+            List<ForeignKey<?, ?>> keys = getReferencesFrom(table, tables);
+            for (ForeignKey<?, ?> key : keys) {
```

Comment:
Just curious - why do we iterate over the same set of keys twice? here an on line 330. Also is the intent to just log the error when composite key is present?

Reply:
Pre-existing logic. I felt it best to leave as-is